### PR TITLE
Generalise AnchoredFragment to AnchoredSeq

### LINKED ISSUE #2594: Generalise AnchoredFragment over the anchor
If we parameterise `AnchoredFragment` over the anchor type, we can reuse this data type for the `LedgerDB` and the `HeaderStateHistory`. Then we would only need one EBB-aware, logarithmic "rewind to `Point`" implementation instead of three.

We should minimise the constraint on the elements in the fragment to something like `HasPoint`, because it doesn't always make sense to require the elements and anchor to have a `BlockNo`. Maybe the fragment in the name is too chain-specific, we can rename it to `AnchoredSeq`?

### PR DESCRIPTION
**Pull request by mrBliss**
Fixes #2594.

Generalise `AnchoredFragment` over its measure, anchor, and element, resulting in `AnchoredSeq`.

Reimplement `AnchoredFragment` using `AnchoredSeq`. `AnchoredFragment`'s testsuite covers `AnchoredSeq`.

Reimplement `LedgerDB` using `AnchoredSeq`.

Reimplement `HeaderStateHistory` using `AnchoredSeq`.

No more need to implement lookup/rollback/shifting the anchor 3 times.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs [hunk 1]
```diff
@@ -692,45 +509,6 @@ applyChainUpdates :: HasHeader block
 applyChainUpdates []     c = Just c
 applyChainUpdates (u:us) c = applyChainUpdates us =<< applyChainUpdate u c
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:510]: Fix `CF.` references

File: ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs [hunk 2]
```diff
@@ -976,17 +708,16 @@ intersectionPoint c1 c2 = case c1 `intersect` c2 of
...
-    AnchoredFragment (castAnchor a) (FT.fmap' f ft)
+mapAnchoredFragment ::
+     (HasHeader block1, HasHeader block2, HeaderHash block1 ~ HeaderHash block2)
+  => (block1 -> block2)
+  -> AnchoredFragment block1
+  -> AnchoredFragment block2
+mapAnchoredFragment = map castAnchor
 
 -- | \( O\(n\) \). Variation on 'filterWithStop' without a stop condition.
 filter ::
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:723]: Generalise to `AnchoredSeq`

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 3]
```diff
@@ -133,18 +125,26 @@ import           Ouroboros.Consensus.Util.CBOR (decodeWithOrigin,
...
+      -- | Ledger states
+      ledgerDbCheckpoints :: !(AnchoredSeq (WithOrigin SlotNo) (Checkpoint l) (Checkpoint l))
 
       -- | Ledger DB parameters
     , ledgerDbParams      :: !LedgerDbParams
     }
   deriving (Show, Eq, Generic, NoThunks)
 
+-- | Newtype wrapper around a ledger state @l@ so that we can define a
+-- non-blankt 'Anchorable' instance.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:138]: blanket

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:138]: Mention that it's internal

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 4]
```diff
@@ -713,36 +540,55 @@ ledgerDbSwitch' cfg n bs db =
...
+-- without extra wrapping.
+--
+-- Compared to the new format, the old format encoded some extra data:
+-- * The tip: @WithOrigin (RealPoint blk)@
+-- * The chain length: @Word64@
+--
+-- The decoder for 'ChainSummary' will accept and ignore them. The encoder will
+-- no longer encode them.
+--
+newtype ChainSummary l blk = ChainSummary {
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:558]: Drop this concept, move the comment to the encoder

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 5]
```diff
@@ -355,250 +314,118 @@ applyBlock cfg ap db = case ap of
...
+-- When no ledger state (or anchor) has the given 'Point', 'Nothing' is
+-- returned.
+ledgerDbPast ::
+     (HasHeader blk, IsLedger l, HeaderHash l ~ HeaderHash blk)
+  => Point blk
+  -> LedgerDB l
+  -> Maybe l
+ledgerDbPast pt db
+    | pt == castPoint (getTip db)
+    = Just $ ledgerDbCurrent db
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:364]: Move to `lookupByMeasure`

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 6]
```diff
@@ -355,250 +314,118 @@ applyBlock cfg ap db = case ap of
...
+    | pt == castPoint (getTip (ledgerDbAnchor db))
+    = Just $ ledgerDbAnchor db
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:366]: Handled in `lookupByMeasure`? Move it there.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 7]
```diff
@@ -355,250 +314,118 @@ applyBlock cfg ap db = case ap of
...
+    | otherwise
+    = fmap unCheckpoint $
+        find ((== pt) . castPoint . getTip . unCheckpoint) $
+          AS.lookupByMeasure (pointSlot pt) (ledgerDbCheckpoints db)
+
+-- | Transform the underlying 'AnchoredSeq' using the given functions.
+ledgerDbMap ::
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:373]: bimap

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 8]
```diff
@@ -355,250 +314,118 @@ applyBlock cfg ap db = case ap of
...
-    | tip == ledgerDbTip db
-    = Just (ledgerDbCurrent db)
-    | tip == csTip (ledgerDbAnchor db)
-    = Just (csLedger (ledgerDbAnchor db))
-    | NotOrigin tip' <- pointToWithOriginRealPoint tip
-    = cpState <$> find ((== tip') . cpBlock) (ledgerDbCheckpoints db)
+rollback :: GetTip l => Word64 -> LedgerDB l -> Maybe (LedgerDB l)
+rollback n db@LedgerDB{..}
+    | n == 0
+    = Just db
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:421]: Drop this

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 9]
```diff
@@ -713,36 +540,55 @@ ledgerDbSwitch' cfg n bs db =
...
+  -> ChainSummary l blk -> Encoding
+encodeChainSummary encodeLedger (ChainSummary l) = mconcat [
+      Enc.encodeListLen 1
+    , encodeLedger l
     ]
 
-decodeChainSummary :: (forall s. Decoder s l)
-                   -> (forall s. Decoder s (HeaderHash blk))
-                   -> forall s. Decoder s (ChainSummary l blk)
+decodeChainSummary ::
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:577]: Extend `decodeVersion` and use that

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs [hunk 10]
```diff
@@ -80,11 +81,13 @@ getCurrentChain CDB{..} =
   where
     SecurityParam k = configSecurityParam cdbTopLevelConfig
 
-getCurrentLedger :: IOLike m => ChainDbEnv m blk -> STM m (ExtLedgerState blk)
+getCurrentLedger ::
+    (LedgerSupportsProtocol blk, IOLike m)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs:85]: Figure out why?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs [hunk 11]
```diff
@@ -92,7 +92,7 @@ getPastLedger ::
 getPastLedger CDB{..} = LgrDB.getPastState cdbLgrDB
 
 getHeaderStateHistory ::
-     IOLike m
+     (LedgerSupportsProtocol blk, IOLike m)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs:95]: Figure out why

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs [hunk 12]
```diff
@@ -306,16 +305,11 @@ getPastState LgrDB{..} p = do
     return $ LedgerDB.ledgerDbPast p db
 
 getHeaderStateHistory ::
-     IOLike m
+     (IOLike m, IsLedger (LedgerState blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/LgrDB.hs:308]: Figure out why

**Review by edsko — APPROVED** [review]
Reviewed over meet

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs (hunk 11 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs:95]: Removed!

Was caused by the unnecessary `Measured` constraint on `FingerTree.fmap'`. We switched to the constraint-less [`unsafeFmap`](https://github.com/input-output-hk/cardano-prelude/commit/906d5f2531cee5d70960413c7795c2cee8566cb7) for `AnchoredSeq.bimapPreservingMeasure`.

Why exactly `LedgerSupportsProtocol`?
```haskell
ledgerDbBimap ::
     (Anchorable (WithOrigin SlotNo) a b, GetTip l)
  => ..

instance IsLedger (LedgerState blk) => GetTip (ExtLedgerState blk) where

instance ( IsLedger (LedgerState  blk)
         , LedgerSupportsProtocol blk
         )
      => IsLedger (ExtLedgerState blk) where
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs (hunk 10 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs:85]: Reduced to `IsLedger (LedgerState blk)`.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/InMemory.hs [hunk 13]
```diff
@@ -55,17 +57,59 @@ tests = testGroup "InMemory" [
...
+      -- tip: WithOrigin (RealPoint TestBlock)
+    , TkListLen 1
+    , TkListLen 2
+    , TkInt 3
+    , TkListBegin, TkInt 0, TkInt 0, TkBreak
+      -- chain length: Word64
+    , TkInt 10
+      -- ledger: Int for simplicity
+    , TkInt 100
+    ]
```
Comment (mrBliss) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/InMemory.hs:93]: I know it's just a unit test, but now I at least now for sure the decoder actually is backwards compatible.

**Review by edsko — APPROVED** [review]
Very nice clean-up, I like it :)

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**